engine.py

```diff
@@ -52,24 +52,80 @@ def get_pawn_moves(self, r, c, moves):
 			return parameter(s):
 			None
 		"""
+		
+		def pawn_attack(pawn_type):
+			"""
+			The pawn attack in diagonal position against opponent piece
+	
+			"""
+			if pawn_type == 'l':
+				if c == 0 and self.board[r-1][c+1][1] == 'd':
+					return moves.append(Move((r, c), (r-1,c+1), self.board))			
+				elif c == 7 and self.board[r-1][c-1][1] == 'd':
+					return moves.append(Move((r, c), (r-1,c-1), self.board))
+				elif c not in [0,7]:
+					for i in range(-1, 2, 2):  # loops twice so j == -1 or 1 
+						if self.board[r - 1][c + i][1] == "d":  # if there's a dark piece is diagonal to it
+							return moves.append(Move((r, c), (r-1, c+i), self.board))  # create  move object and append to list
+					
+			elif pawn_type == 'd':
+				if c == 0 and self.board[r+1][c+1][1] == 'l':
+					return moves.append(Move((r, c), (r+1, c+1), self.board))
+				elif c == 7 and self.board[r+1][c-1][1] == 'l':
+					return moves.append(Move((r, c), (r+1, c-1), self.board))
+				elif c not in [0,7]:
+					for i in range(-1, 2, 2):  # loops twice so j == -1 or 1 
+						if self.board[r+1][c+i][1] == 'l':  # if there's a dark piece is diagonal to it
+							return moves.append(Move((r, c), (r+1, c+i), self.board))  # create  move object and append to list			
+			
 
 		## FIX
 		if self.light_to_move: # if it's light's turn to move
 			
 			for i in range(len(self.board)):
 				for j in range(len(self.board[i])):
-					if self.board[i][j] == "  " or self.board[i][j][1] == "d": # if square is empty or square has opponent's piece
-						moves.append(Move((r, c), (i, j), self.board)) # create a move object and append to list
-
-		
+					# first move
+					if r == 6:
+						for k in range(1,3):
+							if self.board[r-k][c] != "  ": # if blocked
+								pawn_attack('l')
+								break
+							else:
+								moves.append(Move((r, c), (r-k, c), self.board))
+								pawn_attack('l')
+
+					# subsequent move			
+					else:
+						if self.board[r-1][c] != "  ": #if blocked
+							pawn_attack('l')
+							break
+						else:
+							moves.append(Move((r, c), (r-1, c), self.board))
+							pawn_attack('l')
 
 		##FIX
 		else: # if it's dark's turn to move
 
 			for i in range(len(self.board)):
 				for j in range(len(self.board[i])):
-					if self.board[i][j] == "  " or self.board[i][j][1] == "l": # if square is empty or square has opponent's piece
-						moves.append(Move((r, c), (i, j), (self.board))) # create a move object and append to moves
+					# first move
+					if r == 1:
+						for k in range(1,3):
+							if self.board[r+k][c] != "  ": # if blocked
+								pawn_attack('d')
+								break
+							else:
+								moves.append(Move((r, c), (r+k, c), self.board))
+								pawn_attack('d')
+
+					# subsequent move			
+					else:
+						if self.board[r+1][c] != "  ": # if blocked
```

Comment:
This line throws an ` IndexError: list index out of range` when dark pawn is on rank 1 (row 7)
please ensure all potential moves are within the size of the board

Reply:
it is fixed now. Thanks